### PR TITLE
dynamic-user: permit the case static uid and gid are different

### LINKED ISSUE #7013: core: DynamicUser=yes with User= fails if static uid and gid are different
### Submission type

 - Bug report

### systemd version the issue has been seen with

235 with PR #7003

### Used distribution

Fedora 26 x86_64

### In case of bug report: Expected behaviour you didn't see

`DynamicUser=yes` with `User=` works fine.

### In case of bug report: Unexpected behaviour you saw

`DynamicUser=yes` with `User=` fails if static uid and gid are different.
```
$ id systemd-timesync
uid=989(systemd-timesync) gid=987(systemd-timesync) groups=987(systemd-timesync)
$ sudo systemctl start systemd-timesyncd.service
$ journalctl -b --no-hostname
(snip)
Oct 06 17:25:49 systemd[1]: Starting Network Time Synchronization...
Oct 06 17:25:49 systemd[2066]: systemd-timesyncd.service: Failed to update dynamic user credentials: Invalid or incomplete multibyte or wide character
Oct 06 17:25:49 systemd[2066]: systemd-timesyncd.service: Failed at step USER spawning /usr/lib/systemd/systemd-timesyncd: Invalid or incomplete multibyte or wide character
Oct 06 17:25:49 systemd[1]: systemd-timesyncd.service: Main process exited, code=exited, status=217/USER
Oct 06 17:25:49 systemd[1]: systemd-timesyncd.service: Failed with result 'exit-code'.
Oct 06 17:25:49 systemd[1]: Failed to start Network Time Synchronization.
(snip)
```
This is caused by https://github.com/systemd/systemd/blob/master/src/core/dynamic-user.c#L471-L473

### PR DESCRIPTION
**Pull request by yuwata**
This makes systemd supports the case that `DynamicUser=yes` and static user and group exist such that uid and gid of them are different. We only refuse the operation when user does not exist but the group with the same name exists.

Fixes #7013.

### CODE REVIEW

File: src/core/dynamic-user.c [hunk 1]
```diff
@@ -460,19 +460,28 @@ int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *ret) {
...
+                        if (is_user) {
+                                /* OK, this is not a numeric UID. Let's see if there's a user by this name */
+                                p = getpwnam(d->name);
+                                if (p)
+                                        uid = p->pw_uid;
+                                else {
+                                        /* if the user does not exist but the group with the same name exists, refuse operation */
+                                        g = getgrnam(d->name);
+                                        if (g)
+                                                return -EILSEQ;
```
Comment (yuwata) [inline src/core/dynamic-user.c:472]: I am not sure whether the error code is suitable or not. If all you have a suggestion, please give me comments.

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
CI failure seems not related. (Maybe related to #7003.)
```
-------- journal for failed service systemd-timesyncd.service -----------
-- Logs begin at Wed 2017-10-11 07:01:56 UTC, end at Wed 2017-10-11 07:02:22 UTC. --
Oct 11 07:01:57 autopkgtest systemd-timesyncd[504]: Failed to create state directory: File exists
Oct 11 07:01:57 autopkgtest systemd[1]: systemd-timesyncd.service: Main process exited, code=exited, status=1/FAILURE
Oct 11 07:01:57 autopkgtest systemd[1]: systemd-timesyncd.service: Failed with result 'exit-code'.
Oct 11 07:01:57 autopkgtest systemd[1]: Failed to start Network Time Synchronization.
Oct 11 07:01:57 autopkgtest systemd[1]: systemd-timesyncd.service: Service has no hold-off time, scheduling restart.
Oct 11 07:01:57 autopkgtest systemd[1]: systemd-timesyncd.service: Scheduled restart job, restart counter is at 2.
Oct 11 07:01:57 autopkgtest systemd[1]: Stopped Network Time Synchronization.
Oct 11 07:01:57 autopkgtest systemd[1]: Starting Network Time Synchronization...
Oct 11 07:01:57 autopkgtest systemd-timesyncd[527]: Failed to create state directory: File exists
```

File: src/core/dynamic-user.c (hunk 1 above)
Comment (keszybz) [reply src/core/dynamic-user.c:472]: The code itself is OK, but the error message will be incomprehensible to users:
> Failed to update dynamic user credentials: Invalid or incomplete multibyte or wide character

Please add a custom error message for this error code in execute.c where dynamic_creds_realize is called.

Comment (yuwata) [reply src/core/dynamic-user.c:472]: Thank you for your comment. I've added commits. Please take a look.

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -206,7 +206,10 @@
         enabled for a unit, the name of the dynamic user/group is implicitly derived from the unit name. If the unit
         name without the type suffix qualifies as valid user name it is used directly, otherwise a name incorporating a
         hash of it is used. If a statically allocated user or group of the configured name already exists, it is used
-        and no dynamic user/group is allocated. Dynamic users/groups are allocated from the UID/GID range
+        and no dynamic user/group is allocated. Note that if <varname>User=</varname> is specified and the static group
+        with the name exists, then it is required that the static user with the name already exists. Similary,
```
Comment (keszybz) [inline man/systemd.exec.xml]: Similar*ly*

Comment (yuwata) [reply man/systemd.exec.xml]: Thanks. Updated.

**Pull request merged**